plugins/MauticCrmBundle/Integration/SalesforceIntegration.php

```diff
@@ -2718,6 +2721,78 @@ public function getFieldsForQuery($object)
 
         return $fields;
     }
+
+    /**
+     * Update the record in each system taking the last modified record.
+     *
+     * @param $leadId
+     * @param string $channel
+     * @param string $sfObject
+     * @param array  $sfIds
+     *
+     * @return int
+     *
+     * @throws ApiErrorException
+     */
+    public function getLeadDoNotContactByDate($channel = 'email',  &$sfRecords = [])
+    {
+        $filters = [];
+        $leadIds = [];
+
+        if (empty($sfRecords)) {
+            return;
+        }
+        foreach ($sfRecords as $record) {
+            $sfObject = isset($record['integration_entity']) ? $record['integration_entity'] : 'Lead';
+            if (isset($record['integration_entity_id'])) {
+                $sfIdsToFind[$sfObject][$record['integration_entity_id']] = $record['internal_entity_id'];
+            }
+        }
+        foreach ($sfIdsToFind as $object) {
+            $sfIds = [];
+            foreach ($object as $sfId => $lead) {
+                $sfIds[]        = $sfId;
+                $leadIds[$sfId] = $lead;
+            }
+            if (!empty($sfIds)) {
+                $fieldString = "'".implode("','", $sfIds)."'";
+                //get last modified date for donot contact
+                $historySelect = 'Select Field, '.$sfObject.'Id, CreatedDate, isDeleted, NewValue from '.$sfObject.'History where Field = \'HasOptedOutOfEmail\' and '.$sfObject.'Id IN ('.$fieldString.') ORDER BY CreatedDate DESC';
+                $queryUrl      = $this->getQueryUrl();
+                $historySF     = $this->getApiHelper()->request('query', ['q' => $historySelect], 'GET', false, null, $queryUrl);
+            }
+        }
+
+        foreach ($leadIds as $sfKey => $leadId) {
```

Comment:
This needs to be optimized. The audit log is very large and doing a like query like this for every contact will be expensive. So we should be fetching audit log entries for the group of lead IDs.

Also note that using getEntity here is doing a useless query. All `getAuditLogs()` is using from the entity is `$lead->getId()`. Whenever that's the case, we can avoid a query by simply using `$this->em->getReference('MauticLeadBundle:Lead', $leadId)`. But this will become obsolete once this is optimized to a single query.